server/src/diagnostic-queue.ts

```diff
@@ -27,27 +51,11 @@ export class DiagnosticEventQueue {
             return;
         }
         const { file } = event.body;
-        const diagnostics = this.diagnostics.get(file) || new Map<EventTypes, lsp.Diagnostic[]>();
-        diagnostics.set(kind, event.body.diagnostics.map(toDiagnostic));
-        this.diagnostics.set(file, diagnostics);
-        this.firePublishDiagnostics(file);
-    }
-
-    protected firePublishDiagnostics = debounce((file: string): void => {
-        const uri = pathToUri(file);
-        const diagnostics = this.getDiagnostics(file);
-        this.publicDiagnostics({ uri, diagnostics });
-    }, 50);
-    protected getDiagnostics(file: string): lsp.Diagnostic[] {
-        const diagnostics = this.diagnostics.get(file);
-        if (!diagnostics) {
-            return [];
-        }
-        const result: lsp.Diagnostic[] = [];
-        for (const value of diagnostics.values()) {
-            result.push(...value);
+        let fileDiagnostics = this.diagnostics.get(file);
+        if (!fileDiagnostics) {
+            fileDiagnostics = new FileDiagnostics(pathToUri(file));
+            this.diagnostics.set(file, fileDiagnostics);
         }
-        return result;
+        fileDiagnostics.updateDiagnostics(kind, event.body.diagnostics).then(this.publishDiagnostics);
```

Comment:
Won't it execute `publishDiagnostics` so many times as `updateDiagnostics` is called instead of once? How about passing `publishDiagnostics` to `FileDiagnostics` and using `firePublishDiagnostics` as it was before, i.e. without any async logic.

Reply:
> Won't it execute `publishDiagnostics` as many times as `updateDiagnostics` is called instead of once?

It won't. If you look at `updateDiagnostics`, you can see it only resolves the promise once if multiple calls happen within the debounce window, since the call to `resolve` is within the function passed to `debounce`.

Reply:
ok, although the code will be less tricky (without any async handling) on impl and client sides:
```ts
    constructor(
        readonly uri: string,
        protected publishDiagnostics: (params: lsp.PublishDiagnosticsParams) => void
    ) { }

    updateDiagnostics(kind: EventTypes, diagnostics: tsp.Diagnostic[]): void {
        this.diagnosticsPerKind.set(kind, diagnostics.map(toDiagnostic));
        this.firePublishDiagnostics();
    }
    protected firePublishDiagnostics = debounce((): void => {
        const { uri } = this;
        const diagnostics = this.getDiagnostics();
        this.publishDiagnostics({ uri, diagnostics });
    }, 50);
```